atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/knowledge/DefaultKnownConcludedTransactions.java

```diff
@@ -0,0 +1,100 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.transaction.knowledge;
+
+import com.google.common.collect.ImmutableRangeSet;
+import com.google.common.collect.Range;
+import com.google.common.collect.RangeSet;
+import com.google.common.collect.Sets;
+import com.palantir.common.concurrent.CoalescingSupplier;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import java.util.concurrent.atomic.AtomicReference;
+
+@SuppressWarnings("UnstableApiUsage") // RangeSet usage
+public final class DefaultKnownConcludedTransactions implements KnownConcludedTransactions {
+    private static final SafeLogger log = SafeLoggerFactory.get(DefaultKnownConcludedTransactions.class);
+    private static final int MAX_ATTEMPTS = 20;
+
+    private final KnownConcludedTransactionsStore knownConcludedTransactionsStore;
+
+    /**
+     * Concurrency: All updates go through {@link #ensureRangesCached(RangeSet)}} and perform CASes to atomically
+     * evolve the value here. Copy on write should be acceptable given these range-sets are not expected to be large.
+     */
+    private final AtomicReference<ImmutableRangeSet<Long>> cachedConcludedTimestamps;
+
+    private final CoalescingSupplier<Void> cacheUpdater = new CoalescingSupplier<>(() -> {
+        updateCacheFromRemote();
+        return null;
+    });
+
+    public DefaultKnownConcludedTransactions(KnownConcludedTransactionsStore knownConcludedTransactionsStore) {
+        this.knownConcludedTransactionsStore = knownConcludedTransactionsStore;
+        this.cachedConcludedTimestamps = new AtomicReference<>(ImmutableRangeSet.of());
+    }
+
+    @Override
+    public boolean isKnownConcluded(long startTimestamp, Consistency consistency) {
+        if (cachedConcludedTimestamps.get().contains(startTimestamp)) {
+            return true;
+        }
+        if (consistency == Consistency.REMOTE_READ) {
+            return performRemoteReadAndCheckConcluded(startTimestamp);
+        }
+        return false;
+    }
+
+    @Override
+    public void addConcludedTimestamps(Range<Long> knownConcludedInterval) {
+        knownConcludedTransactionsStore.supplement(knownConcludedInterval);
+        ensureRangesCached(ImmutableRangeSet.of(knownConcludedInterval));
```

Comment:
we could just use `updateCacheFromRemote` - we'll probably be able to coalesce this request with those coming from performRemoteReadAndCheckConcluded and `updateCacheFromRemote` will  at least include `knownConcludedInterval`

Reply:
Let's discuss: there are some tradeoffs here. In particular, `ensureRangesCached()` does not perform a database read!

Reply:
oh right i see. Okay in that case probably this is better - I might have misread the atomicRef get and thought it was the coalescing supplier get()